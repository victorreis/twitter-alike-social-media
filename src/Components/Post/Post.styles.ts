import styled, { CSSObject, DefaultTheme } from 'styled-components';

import { Typography } from '../Typography';

const getPostBorders = ({
  theme,
  compact,
}: {
  theme: DefaultTheme;
  compact: boolean;
}): CSSObject => {
  if (compact) {
    return {
      borderWidth: '1px',
      borderStyle: 'solid',
      borderColor: theme.colors.background.default.light,
      borderRadius: 16,
    };
  }

  return {
    borderBottomWidth: '1px',
    borderBottomStyle: 'solid',
    borderBottomColor: theme.colors.background.default.normal,
  };
};

const getPostContentButtonsContainerBorder = ({
  theme,
  hasChildren,
}: {
  theme: DefaultTheme;
  hasChildren: boolean;
}): CSSObject => {
  return {
    ...(!hasChildren && {
      borderTopWidth: '1px',
      borderTopStyle: 'solid',
      borderTopColor: theme.colors.background.default.normal,
    }),
  };
};

export const PostContainer = styled.div`
  display: flex;
  padding: 16px;
  column-gap: 1rem;

  ${getPostBorders}
`;

export const PostContent = styled.div`
  display: flex;
  flex-direction: column;
  column-gap: 1rem;
  flex: 1;
`;

export const PostContentHeader = styled.div`
  display: flex;
  column-gap: 1rem;
  align-items: center;
`;

export const PostContentTitle = styled(Typography)`
  margin: 0;
`;

export const PostText = styled(Typography)`
  flex: 1;
  margin-bottom: 16px;
`;

export const PostContentButtonsContainer = styled.div<{ hasChildren: boolean }>`
  display: flex;
  flex: 1;
  column-gap: 1rem;
  align-items: center;
  align-self: stretch;
  justify-content: flex-end;
  padding-top: 16px;

  ${getPostContentButtonsContainerBorder};
`;
