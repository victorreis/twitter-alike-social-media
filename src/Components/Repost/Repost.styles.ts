import { AiOutlineRetweet } from 'react-icons/ai';

import styled from 'styled-components';

import { Typography } from '../Typography';

export const RepostContainer = styled.div`
  display: flex;
  flex-direction: column;
`;

export const RepostTextContainer = styled.div`
  display: flex;
  column-gap: 0.5rem;
  padding: 16px;
  padding-bottom: 0;
`;

export const RepostText = styled(Typography)`
  margin: 0;
  color: ${({ theme }) => theme.colors.font.hint};
`;

export const RepostIcon = styled(AiOutlineRetweet)`
  color: ${({ theme }) => theme.colors.font.hint};
`;
