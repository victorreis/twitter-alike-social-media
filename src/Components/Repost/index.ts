export * from './Repost.types';
export * from './Repost';
