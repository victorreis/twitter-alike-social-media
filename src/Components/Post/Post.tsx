import { useMemo, useRef, useState } from 'react';

import dayjs from 'dayjs';
import { nanoid } from 'nanoid';

import { EXTENDED_DATE_FORMAT } from '../../Config/Constants.config';
import { TestProps } from '../../Config/Tests/Test.types';
import { TypographyVariant } from '../../Theme/Types/Typographies.types';
import { Avatar } from '../Avatar';
import { Button, ButtonSize } from '../Button';
import { PostCreator } from '../PostCreator';
import { Typography } from '../Typography';
import {
  PostContainer,
  PostContent,
  PostContentButtonsContainer,
  PostContentHeader,
  PostContentTitle,
  PostText,
} from './Post.styles';
import { PostProps, DefaultPostProps } from './Post.types';

export const postDefaults: Required<DefaultPostProps> & Required<TestProps> = {
  testID: 'Post',
  compact: false,
};

export const Post: React.FC<PostProps> = (props): JSX.Element => {
  const {
    testID = postDefaults.testID,
    children,
    id,
    text,
    createdAt,
    createdBy,
    reposts,
    quotPosts,
    compact = postDefaults.compact,
    isAvatarClickable,
    onClickAvatar,
    onClickRepost,
    onClickQuotePost,
    style,
    ...others
  } = props;

  const { name, nickname, thumbnailUrl } = createdBy;
  const [showPostCreator, setShowPostCreator] = useState<boolean>();
  const quotePostCreatorRef = useRef<HTMLTextAreaElement>(null);

  const formattedDate = dayjs(createdAt).format(EXTENDED_DATE_FORMAT);

  const userNameVariant: TypographyVariant = compact ? 'h5' : 'h4';
  const userInfoVariant: TypographyVariant = compact
    ? 'subtitle3'
    : 'subtitle2';
  const textVariant: TypographyVariant = compact ? 'body2' : 'body1';
  const buttonSize: ButtonSize = compact ? 'MD' : 'LG';

  const handleClickRepost = () => {
    onClickRepost(id);
  };

  const handleSubmitQuotePost = (quotePostText: string) => {
    setShowPostCreator(() => false);
    onClickQuotePost(id, quotePostText);
  };

  const renderText = useMemo(
    () => () => {
      return text.split('\n').map((line) => (
        <span key={nanoid()}>
          {line}
          <br />
        </span>
      ));
    },
    [text]
  );

  const renderAvatar = () => {
    return (
      <Avatar
        isAvatarClickable={Boolean(isAvatarClickable)}
        name={name}
        nickname={nickname}
        onClickAvatar={onClickAvatar}
        size={compact ? 'SM' : 'MD'}
        thumbnailUrl={thumbnailUrl}
      />
    );
  };

  const handleShowQuotePostCreator = () => {
    setShowPostCreator(() => true);
    if (quotePostCreatorRef.current) {
      quotePostCreatorRef.current.focus();
    }
  };

  const renderQuotePostCreator = () => {
    return (
      showPostCreator && (
        <PostCreator
          ref={quotePostCreatorRef}
          isAvatarClickable={isAvatarClickable}
          name={name}
          nickname={nickname}
          onClickAvatar={onClickAvatar}
          onSubmit={handleSubmitQuotePost}
          thumbnailUrl={thumbnailUrl}
        />
      )
    );
  };

  return (
    <PostContainer
      key={id}
      compact={compact}
      data-testid={testID}
      style={style}
      {...others}
    >
      {!compact && renderAvatar()}

      <PostContent>
        <PostContentHeader>
          {compact && renderAvatar()}

          <PostContentTitle variant={userNameVariant}>{name}</PostContentTitle>
          <PostContentTitle variant={userInfoVariant}>
            @{nickname}
          </PostContentTitle>
          <Typography variant={userInfoVariant}>- {formattedDate}</Typography>
        </PostContentHeader>

        <PostText data-testid={`${testID}_text`} variant={textVariant}>
          {renderText()}
        </PostText>

        {children}

        <PostContentButtonsContainer hasChildren={Boolean(children)}>
          <Button onClick={handleClickRepost} size={buttonSize}>
            Repost
          </Button>
          <Typography variant={textVariant}>{reposts}</Typography>
          <Button
            disabled={showPostCreator}
            onClick={handleShowQuotePostCreator}
            size={buttonSize}
          >
            Quote post
          </Button>

          <Typography variant={textVariant}>{quotPosts}</Typography>
        </PostContentButtonsContainer>

        {renderQuotePostCreator()}
      </PostContent>
    </PostContainer>
  );
};
