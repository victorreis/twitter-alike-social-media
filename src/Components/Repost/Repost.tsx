import { TestProps } from '../../Config/Tests/Test.types';
import { Post } from '../Post';
import {
  RepostContainer,
  RepostIcon,
  RepostText,
  RepostTextContainer,
} from './Repost.styles';
import { RepostProps, DefaultRepostProps } from './Repost.types';

export const repostDefaults: Required<DefaultRepostProps> &
  Required<TestProps> = {
  testID: 'Repost',
};

export const Repost: React.FC<RepostProps> = (props): JSX.Element => {
  const {
    testID = repostDefaults.testID,
    style,
    originalPost,
    createdBy,
    onClickRepost,
    onClickQuotePost,
    ...others
  } = props;

  const handleClickRepostOriginalPost = () => {
    onClickRepost(originalPost.id);
  };

  const handleClickQuotePostOriginalPost = () => {
    onClickQuotePost(originalPost.id);
  };

  return (
    <RepostContainer data-testid={testID} style={style} {...others}>
      <RepostTextContainer>
        <RepostIcon size={20} />
        <RepostText variant="h5">{createdBy.name} Reposted</RepostText>
      </RepostTextContainer>

      <Post
        {...originalPost}
        onClickQuotePost={handleClickQuotePostOriginalPost}
        onClickRepost={handleClickRepostOriginalPost}
      />
    </RepostContainer>
  );
};
