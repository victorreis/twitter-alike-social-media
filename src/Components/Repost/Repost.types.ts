import { TestProps } from '../../Config/Tests/Test.types';
import { RepostType } from '../../Models/Repost.types';
import { AvatarClickProps } from '../Avatar';
import { RequiredPostProps } from '../Post';

export interface RequiredRepostProps {}

export interface DefaultRepostProps {}

export interface OptionalRepostProps {}

export type RepostProps = RequiredRepostProps &
  DefaultRepostProps &
  OptionalRepostProps &
  TestProps &
  RepostType &
  RequiredPostProps &
  AvatarClickProps &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;

export type RepostStyleProps = Required<DefaultRepostProps>;
