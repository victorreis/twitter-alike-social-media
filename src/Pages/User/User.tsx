import { useState, useEffect } from 'react';

import { PageContainer } from '../../App.styles';
import { PostCreator } from '../../Components/PostCreator';
import { Tab } from '../../Components/Tab';
import { UserDetails } from '../../Components/UserDetails';
import { TestProps } from '../../Config/Tests/Test.types';
import { useLoggedUser } from '../../Hooks/LoggedUser';
import { useRenderPosts } from '../../Hooks/RenderPosts';
import { useShowUserPage } from '../../Hooks/ShowUserPage';
import { UserType } from '../../Models/User.types';
import { followerFollowedCreatorService } from '../../Services/FollowerFollowedCreator';
import { FollowerFollowedRetrieverService } from '../../Services/FollowerFollowedRetriever';
import { PostTypes } from '../../Services/LocalStorageInitializer';
import { postRetrieverService } from '../../Services/PostRetriever';
import { userRetrieverService } from '../../Services/UserRetriever';
import { UserTabBar } from './User.styles';

export const userDefaults: Required<TestProps> = {
  testID: 'User',
};

export const User: React.FC = (): JSX.Element | null => {
  const [user, setUser] = useState<UserType>();
  const [posts, setPosts] = useState<PostTypes[]>([]);
  const [isFollowing, setIsFollowing] = useState<boolean | undefined>(
    undefined
  );
  const { renderPosts } = useRenderPosts({ posts });

  const { loggedUser } = useLoggedUser();
  const { urlNickname, handleCloseUserPage } = useShowUserPage({
    nickname: user?.nickname,
  });

  useEffect(() => {
    const loadedUser = userRetrieverService.getByNickname(String(urlNickname));
    if (!loadedUser) {
      handleCloseUserPage();
      return;
    }
    setUser(() => loadedUser);

    const loadedPosts = postRetrieverService.getAllCreatedByUser(loadedUser.id);
    setPosts(() => loadedPosts);

    if (loggedUser && loggedUser.id !== loadedUser.id) {
      setIsFollowing(() =>
        FollowerFollowedRetrieverService.isFollowerFollowed(
          loggedUser.id,
          loadedUser.id
        )
      );
    }
  }, [handleCloseUserPage, loggedUser, loggedUser?.id, urlNickname]);

  const handleFollow = () => {
    if (loggedUser && user) {
      followerFollowedCreatorService.createRelationship(loggedUser.id, user.id);
      setIsFollowing((prevState) => !prevState);
    }
  };

  const handleUnfollow = () => {
    if (loggedUser && user) {
      followerFollowedCreatorService.deleteRelationship(loggedUser.id, user.id);
      setIsFollowing((prevState) => !prevState);
    }
  };

  if (!user) {
    return null;
  }

  const { name, thumbnailUrl } = user;

  return (
    <PageContainer data-testid={userDefaults.testID}>
      <UserDetails
        {...user}
        isFollowing={isFollowing}
        onFollow={handleFollow}
        onUnfollow={handleUnfollow}
      />

      <PostCreator
        name={name}
        nickname={String(urlNickname)}
        onChange={() => {}}
        thumbnailUrl={thumbnailUrl}
      />

      <UserTabBar>
        <Tab active text="POSTS" />
      </UserTabBar>

      {renderPosts()}
    </PageContainer>
  );
};
