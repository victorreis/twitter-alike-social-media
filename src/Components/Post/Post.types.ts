import { TestProps } from '../../Config/Tests/Test.types';
import { PostType } from '../../Post.types';
import { UserType } from '../../User.types';

export interface RequiredPostProps {
  /**
   * User that created the post.
   */
  createdBy: UserType;
}

export interface DefaultPostProps {
  /**
   * Controls when the Post is compact or not.
   */
  compact?: boolean;
}

export interface OptionalPostProps {
  /**
   * Sets the component styles.
   */
  style?: React.CSSProperties;
}

export type PostProps = RequiredPostProps &
  DefaultPostProps &
  OptionalPostProps &
  TestProps &
  Omit<PostType, 'createdBy'> &
  Omit<React.HTMLAttributes<HTMLSpanElement>, 'children'>;
