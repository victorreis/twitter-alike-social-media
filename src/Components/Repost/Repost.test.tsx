import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { PostType } from '../../Post.types';
import { RepostType } from '../../Repost.types';
import { UserType } from '../../User.types';
import { postDefaults } from '../Post';
import { Repost, repostDefaults } from './Repost';
import { RepostProps } from './Repost.types';

describe('repost component tests', () => {
  const userTeste: UserType = {
    id: 'user123',
    name: 'Victor Reis',
    nickname: 'victor.reis',
    url: '...',
    thumbnailUrl: 'https://i.pravatar.cc/300',
    createdAt: new Date(),
    following: 333,
    followers: 777,
  };
  const post: PostType = {
    id: 'post123',
    text: `first line\nsecond line\nthird line`,
    createdAt: new Date(),
    createdBy: userTeste,
    reposts: 9,
    quotPosts: 99,
  };
  const repost: RepostType = {
    id: 'post123',
    createdAt: new Date(),
    createdBy: userTeste,
    originalPost: post,
  };

  const requiredProps: RepostProps = {
    ...repost,
  };

  const setup = (props?: RepostProps) => {
    const renderRTR = () =>
      renderRTRCreator<RepostProps>(Repost, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<RepostProps>(Repost, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(repostDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the Post component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(postDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the text`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const element = screen.getByText(`${repost.createdBy.name} Reposted`);

      expect(element).toBeInTheDocument();
    });
  });

  describe('style tests', () => {});

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
