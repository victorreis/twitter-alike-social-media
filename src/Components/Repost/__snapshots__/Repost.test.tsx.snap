// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repost component tests snapshot tests should render correctly 1`] = `
<div
  className="RepostContainer-twitter-alike-social-media__sc-dgb1yt-0 efDaFT"
  data-testid="Repost"
  id="post123"
  type="Repost"
>
  <div
    className="RepostTextContainer-twitter-alike-social-media__sc-dgb1yt-1 hxLbHP"
  >
    <svg
      className="RepostIcon-twitter-alike-social-media__sc-dgb1yt-3 bStOLt"
      fill="currentColor"
      height={20}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width={20}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M136 552h63.6c4.4 0 8-3.6 8-8V288.7h528.6v72.6c0 1.9.6 3.7 1.8 5.2a8.3 8.3 0 0 0 11.7 1.4L893 255.4c4.3-5 3.6-10.3 0-13.2L749.7 129.8a8.22 8.22 0 0 0-5.2-1.8c-4.6 0-8.4 3.8-8.4 8.4V209H199.7c-39.5 0-71.7 32.2-71.7 71.8V544c0 4.4 3.6 8 8 8zm752-80h-63.6c-4.4 0-8 3.6-8 8v255.3H287.8v-72.6c0-1.9-.6-3.7-1.8-5.2a8.3 8.3 0 0 0-11.7-1.4L131 768.6c-4.3 5-3.6 10.3 0 13.2l143.3 112.4c1.5 1.2 3.3 1.8 5.2 1.8 4.6 0 8.4-3.8 8.4-8.4V815h536.6c39.5 0 71.7-32.2 71.7-71.8V480c-.2-4.4-3.8-8-8.2-8z"
      />
    </svg>
    <h5
      className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 cCEHbN RepostText-twitter-alike-social-media__sc-dgb1yt-2 epWPqF"
      data-testid="Typography"
    >
      Victor Reis
       Reposted
    </h5>
  </div>
  <div
    className="PostContainer-twitter-alike-social-media__sc-zy8jam-0 gsEoCj"
    data-testid="Post"
    type="Post"
  >
    <div
      className="AvatarContainer-twitter-alike-social-media__sc-ne4tqs-0 czdfxE"
      data-testid="Avatar"
      onClick={[Function]}
      size="MD"
    >
      <img
        alt="Victor Reis's avatar photo."
        className="AvatarImage-twitter-alike-social-media__sc-ne4tqs-1 dnbueO"
        data-testid="Avatar_image"
        onError={[Function]}
        size="MD"
        src="https://i.pravatar.cc/300"
        title="Victor Reis's avatar photo."
      />
    </div>
    <div
      className="PostContent-twitter-alike-social-media__sc-zy8jam-1 kvzPWr"
    >
      <div
        className="PostContentHeader-twitter-alike-social-media__sc-zy8jam-2 ilvDfQ"
      >
        <h4
          className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 dwIaAW PostContentTitle-twitter-alike-social-media__sc-zy8jam-3 bFlsZm"
          data-testid="Typography"
        >
          Victor Reis
        </h4>
        <h4
          className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 cCEHbN PostContentTitle-twitter-alike-social-media__sc-zy8jam-3 bFlsZm"
          data-testid="Typography"
        >
          @
          victor.reis
        </h4>
        <h4
          className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 cCEHbN"
          data-testid="Typography"
        >
          - 
          January 1, 2011
        </h4>
      </div>
      <span
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT PostText-twitter-alike-social-media__sc-zy8jam-4 kKAkLD"
        data-testid="Post_text"
      >
        <span>
          first line
          <br />
        </span>
        <span>
          second line
          <br />
        </span>
        <span>
          third line
          <br />
        </span>
      </span>
      <div
        className="PostContentButtonsContainer-twitter-alike-social-media__sc-zy8jam-5 gsSVXy"
      >
        <button
          className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 fBnyBV"
          data-testid="Button"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          size="LG"
          type="button"
        >
          Repost
        </button>
        <span
          className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
          data-testid="Typography"
        >
          9
        </span>
        <button
          className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 fBnyBV"
          data-testid="Button"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          size="LG"
          type="button"
        >
          Quote post
        </button>
        <span
          className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
          data-testid="Typography"
        >
          99
        </span>
      </div>
    </div>
  </div>
</div>
`;
